src/seq_mv/csr_matop.c

```diff
@@ -14,6 +14,325 @@
+         {
+            hypre_assert(C_i[iic+1] == 0);
+            C_i[iic+1] = C_i[rownnz_C[lastrow-1]+1];
+         }
+      }
+   }
+
+#ifdef HYPRE_USING_OPENMP
+#pragma omp barrier
+#endif
+
+#ifdef HYPRE_DEBUG
+   if (!ii)
+   {
+      for (i = 0; i < nrows_C; i++)
+      {
+         hypre_assert(C_i[i] <= C_i[i+1]);
+         hypre_assert(((A_i[i+1] - A_i[i]) +
+                       (B_i[i+1] - B_i[i])) >=
+                       (C_i[i+1] - C_i[i]));
+         hypre_assert((C_i[i+1] - C_i[i]) >= (A_i[i+1] - A_i[i]));
+         hypre_assert((C_i[i+1] - C_i[i]) >= (B_i[i+1] - B_i[i]));
+      }
+      hypre_assert((C_i[nrows_C] - C_i[0]) == num_nonzeros);
+   }
+#endif
+
+   return hypre_error_flag;
+}
+
+/*--------------------------------------------------------------------------
+ * hypre_CSRMatrixAddSecondPass:
+ *
+ * Performs the second pass needed for Matrix/Matrix addition (C = A + B).
+ * This function computes C_j and C_data.
+ *
+ * Notes: see notes for hypre_CSRMatrixAddFirstPass
+ *--------------------------------------------------------------------------*/
+HYPRE_Int
+hypre_CSRMatrixAddSecondPass( HYPRE_Int          firstrow,
+                              HYPRE_Int          lastrow,
+                              HYPRE_Int         *twspace,
+                              HYPRE_Int         *marker,
+                              HYPRE_Int         *map_A2C,
+                              HYPRE_Int         *map_B2C,
+                              HYPRE_Int         *rownnz_C,
+                              HYPRE_Complex      alpha,
+                              HYPRE_Complex      beta,
+                              hypre_CSRMatrix   *A,
+                              hypre_CSRMatrix   *B,
+                              hypre_CSRMatrix   *C )
+{
+   HYPRE_Int        *A_i      = hypre_CSRMatrixI(A);
+   HYPRE_Int        *A_j      = hypre_CSRMatrixJ(A);
+   HYPRE_Complex    *A_data   = hypre_CSRMatrixData(A);
+   HYPRE_Int         nnzs_A   = hypre_CSRMatrixNumNonzeros(A);
+
+   HYPRE_Int        *B_i      = hypre_CSRMatrixI(B);
+   HYPRE_Int        *B_j      = hypre_CSRMatrixJ(B);
+   HYPRE_Complex    *B_data   = hypre_CSRMatrixData(B);
+   HYPRE_Int         nnzs_B   = hypre_CSRMatrixNumNonzeros(A);
+
+   HYPRE_Int        *C_i      = hypre_CSRMatrixI(C);
+   HYPRE_Int        *C_j      = hypre_CSRMatrixJ(C);
+   HYPRE_Complex    *C_data   = hypre_CSRMatrixData(C);
+   HYPRE_Int         ncols_C  = hypre_CSRMatrixNumCols(C);
+
+   HYPRE_Int         ia, ib, ic, iic;
+   HYPRE_Int         jcol, pos;
+
+   hypre_assert(( map_A2C &&  map_B2C) ||
+                (!map_A2C && !map_B2C) ||
+                ( map_A2C && (nnzs_B == 0)) ||
+                ( map_B2C && (nnzs_A == 0)));
+
+   /* Initialize marker vector */
+   for (ia = 0; ia < ncols_C; ia++)
+   {
+      marker[ia] = -1;
+   }
+
+   pos = C_i[rownnz_C ? rownnz_C[firstrow] : firstrow];
+   if ((map_A2C && map_B2C) || ( map_A2C && (nnzs_B == 0)) || ( map_B2C && (nnzs_A == 0)))
```

Comment:
Can we combine the if-branch with the else-, having `jcol = map_A2C? map_A2C[A_j[ia]] : A_j[ia]` and similar for B? Just to simplify.

Reply:
I thought about doing this and decided to do the other way. The problem with `jcol = map_A2C? map_A2C[A_j[ia]] : A_j[ia]` is that the conditional statement is evaluated `num_nonzero_A` times and this can be quite big. In the way I'm currently doing, the conditional statements on `map_A2C` and `map_B2C` are evaluated only once. Do you think the more compact form should be used? I'm happy to change it if so

Reply:
OK. Thanks for explaining. I think we should keep the current form. The cost of  the conditional statements could be non-negligible. Thanks!